File: Hop_Field/Hop_Field.py
#!/usr/bin/python3

import numpy as np

class HopFieldNetwork() :
    def __init__(self, N) :
        self.N = N
        self.weight_matrix = np.zeros((N, N))
        self.isNormalised = False

    def learn_input(self, X) :
        X = np.array(X)
        X = X.reshape(self.N, 1)
        X_transpose = X.transpose()
        self.weight_matrix += X.dot(X_transpose)
        self.make_diag_zero()

    def make_diag_zero(self) :
        for i in range(self.N) :
            self.weight_matrix[i, i] = 0

    def normalise(self) :
        if not self.isNormalised :
            self.weight_matrix /= self.N
            self.isNormalised = True

    def predict_one_val(self, X, i) :
        self.normalise()
        X = np.array(X)
        i_th_column_of_weight_matrix = self.weight_matrix[:, i]
        net = X.dot(i_th_column_of_weight_matrix)
        val = X[i]
        X[i] = self.activation(net, val)

        return X

    def predict_one_step(self, X) :
        # print("LENGHT: ", len(X), self.N)
        for i in range(self.N) :
            X = self.predict_one_val(X, i)
        return X
    
    def predict(self, X) :
        prev_inp = None
        self.steps = 0
        # print("HAHA: ", len(X), self.N)
        while True :
            X = self.predict_one_step(X)
            self.steps += 1
            if all(prev_inp == X) : # All elements of two arrays are same
                return X
            prev_inp = X

    def activation(self, net, val) :
        if net > 0 :
            return 1
        elif net == 0 :
            return val
        else :
            return -1

File: Hop_Field/Hop_Field.js
class HopFieldNetwork{
    constructor(N){
        this.N = N
        this.weight_matrix = new Matrix(N, N)
        this.isNormalized = false
        // this.weight_matrix.display()
    }

    learn_input(X){
        X = Matrix.arrayToMatrix(X)
        var X_transpose = X.transpose()
        var dot_product = X_transpose.multiply(X)
        this.weight_matrix.add(dot_product)
        this.weight_matrix.display()
        this.make_diagonal_zero()
    }

    make_diagonal_zero(){
        for(var i = 0; i < this.N; i++){
            this.weight_matrix.table[i, i] = 0
        }
    }

    normalise(){
        if(! this.isNormalized){
            for(var i = 0; i < this.weight_matrix.table.rows; i++){
                for(var j = 0; j < this.weight_matrix.table.cols; j++){
                    this.weight_matrix.table[i][j] /= N
                }
            }
            this.isNormalized = true
        }
    }

    predict_one_val(X, i){
        this.normalise()
        X = Matrix.arrayToMatrix(X)
        ith_column_of_weight_matrix = this.weight_matrix.get_column(i)
        net = X.multiply(ith_column_of_weight_matrix)
        val = X.table[i]
        X.table[i] = this.activation(net, val)
        self.steps += 1
        
        return X
    }

    predict_one_step(X){
        for(var i = 0; i < this.N; i++){
            X = predict_one_val(X, i)
        }
        return X
    }

    predict(X){
        prev_inp = null
        this.steps = 0

        while(true){
            X = this.predict_one_step(X)

            if(prev_inp == X){
                return X
            }
            prev_inp = X
        }
    }

    activation(net, val){
        if(net > 0){
            return 1
        }
        else if(net == 0){
            return val
        }
        else{
            return -1
        }
    }

}